Fix writing of last word on DWORD TPI parts

danapplegate opened the pull request:
This PR addresses an issue I discovered last Summer and first reported in [this AVR Freaks forum post](https://www.avrfreaks.net/forum/problems-uploading-only-certain-programs-avrdude-attiny20-using-tpi-ft232h).

While AVRDude's TPI code implements the WORD_WRITE NVMCMD for writing a single word at a time to flash memory, the ATtiny20 (and possibly other parts, I haven't checked) instead specifies a DWORD_WRITE instruction which "conveniently" uses the same NVMCMD hex code. From the [Tiny20's datasheet, section 19.4.3](https://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-8235-8-bit-AVR-Microcontroller-ATtiny20_Datasheet.pdf):

> The Flash is written two words at a time but the data space uses byte-addressing to access Flash that has been mapped to data memory. It is therefore important to write the two words in the correct order to the Flash, namely low bytes before high bytes. The low byte of the first word is first written to the temporary buffer, then the high byte. **Writing the low byte and then the high byte to the buffer latches the two words into the Flash write buffer, starting the actual Flash write operation.**

This means that programs with an odd word count will fail to write the final word. Given this minimalist example program, gcc-avr compiles it to 66 bytes, or 33 words:

```cpp
#ifndef __AVR_ATtiny20__
#define __AVR_ATtiny20__
#endif

#include <avr/cpufunc.h>

int main(void)
{
	_NOP();
	_NOP();
}
```

This causes AVRDude to think it's successfully written the full program, yet it fails when verifying the final two bytes:

```
avrdude: writing 66 bytes flash ...

Writing | ################################################## | 100% 1.16s

avrdude: 66 bytes of flash written
avrdude: verifying flash memory against tpi_dword_test.hex
avrdude: reading on-chip flash data ...

Reading | ################################################## | 100% 1.08s

avrdude: verifying ...
avrdude: verification error, first mismatch at byte 0x0041
         0xff != 0xcf
avrdude: verification error; content mismatch
```

My proposed change to `avr_write_mem()`'s TPI loop is to write a final 0xFFFF word if the total word count is odd. This should fill the second word of the flash write buffer and trigger the write on DWORD_WRITE chips. For single-word chips, my assumption is that writing this extra word won't hurt and is probably fine. However, if needed, I could expand the PR to include a specific DWORD configuration setting that would alter behavior only for specified chips.

With this proposed change, AVRDude successfully writes the final word:

```
avrdude: writing 66 bytes flash ...

Writing | ################################################## | 100% 1.20s

avrdude: 66 bytes of flash written
avrdude: verifying flash memory against tpi_dword_test.hex
avrdude: reading on-chip flash data ...

Reading | ################################################## | 100% 1.08s

avrdude: verifying ...
avrdude: 66 bytes of flash verified
```

MCUdude commented:
Thanks for the PR @danapplegate!

Would you mind compiling your test program for an ATtiny10 as well, and uploading the hex file here? I have a T10-based test board I can use to verify the issue.

danapplegate commented:
@MCUdude Sure, though it looks like [the t10 doesn't use DWORD_WRITE](https://ww1.microchip.com/downloads/aemDocuments/documents/OTH/ProductDocuments/DataSheets/ATtiny4-5-9-10-Data-Sheet-DS40002060A.pdf), so I wouldn't expect the problem to appear with it.

Here's the hex file for the [t10](https://gist.github.com/danapplegate/fa560171f760dc2e778af4fee69331c0).

danapplegate commented:
Taking a look at the [ATtiny wikipedia page](https://en.wikipedia.org/wiki/ATtiny_microcontroller_comparison_chart), it looks like, of all the models that use TPI, the ATtiny4, ATtiny5, ATtiny9, ATtiny10, ATtiny102, and ATtiny104 all use single-word write instructions. ATtiny20 seems to be the only one that uses the double-word instructions.

However, it seems that the ATtiny40 [must have 4 words written at a time](https://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-8263-8-bit-AVR-Microcontroller-tinyAVR-ATtiny40_Datasheet.pdf). So, I would bet that a similar issue would arise when trying to flash programs with a word count that is not divisible by 4 to the t40.

MCUdude commented:
I can confirm that the hex file you uploaded for the ATtiny10 still has an odd number of *words*, but uploads without any issues.

```
$ ./avrdude -cusbasp -pattiny10 -Uflash:w:/Users/hans/Downloads/tpi_dword_test_t10.hex:i

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.00s

avrdude: Device signature = 0x1e9003 (probably t10)
avrdude: NOTE: "flash" memory has been specified, an erase cycle will be performed
         To disable this feature, specify the -D option.
avrdude: erasing chip
avrdude: reading input file /Users/hans/Downloads/tpi_dword_test_t10.hex for flash
avrdude: writing 54 bytes flash ...

Writing | ################################################## | 100% 0.22s

avrdude: 54 bytes of flash written
avrdude: verifying flash memory against /Users/hans/Downloads/tpi_dword_test_t10.hex

Reading | ################################################## | 100% 0.02s

avrdude: 54 bytes of flash verified

avrdude done.  Thank you.
```

> However, it seems that the ATtiny40 [must have 4 words written at a time](https://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-8263-8-bit-AVR-Microcontroller-tinyAVR-ATtiny40_Datasheet.pdf). So, I would bet that a similar issue would arise when trying to flash programs with a word count that is not divisible by 4 to the t40.

Any idea how we can prove/confirm this? I have an ATtiny40 I can test with, if you can provide a hex file you believe may cause issues for the T40.

I tried to upload the T10 hex file to the T40, and this works just fine:

```
$ ./avrdude -cusbasp -pattiny40 -Uflash:w:/Users/hans/Downloads/tpi_dword_test_t10.hex:i

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.00s

avrdude: Device signature = 0x1e920e (probably t40)
avrdude: NOTE: "flash" memory has been specified, an erase cycle will be performed
         To disable this feature, specify the -D option.
avrdude: erasing chip
avrdude: reading input file /Users/hans/Downloads/tpi_dword_test_t10.hex for flash
avrdude: writing 54 bytes flash ...

Writing | ################################################## | 100% 0.12s

avrdude: 54 bytes of flash written
avrdude: verifying flash memory against /Users/hans/Downloads/tpi_dword_test_t10.hex

Reading | ################################################## | 100% 0.02s

avrdude: 54 bytes of flash verified

avrdude done.  Thank you.
```

danapplegate commented:
> I can confirm that the hex file you uploaded for the ATtiny10 still has an odd number of words, but uploads without any issues.

@MCUdude Makes sense. Since the T10's WORD_WRITE NVMCMD only writes a word at a time, the number of total words shouldn't matter.

> Any idea how we can prove/confirm this? I have an ATtiny40 I can test with, if you can provide a hex file you believe may cause issues for the T40.

I have a theory. I see you're using a `usbasp` programmer, which seems to use `paged_write` mode and importantly doesn't set `pgm->cmd_tpi`. That would skip [this block in `avr_write_mem` starting on line 858](https://github.com/avrdudes/avrdude/blob/44fe5bec2dd2f016d135f0d2f02173045d7aeef5/src/avr.c#L858):

```cpp
  if ((p->prog_modes & PM_TPI) && m->page_size > 1 && pgm->cmd_tpi) {
```

Instead, I'm assuming `usbasp` instead relies on `paged_write` mode, which may not exhibit this bug. Perhaps it aligns its writes to page boundaries, which ensures that enough words get written to trigger the flash operation.

I'm using an FTDI chip in MPSSE bit-bang mode, which [does seem to set cmd_tpi](https://github.com/avrdudes/avrdude/blob/main/src/avrftdi_tpi.c#L98), meaning that it enters the block in `avr_write_mem` that I patched in my PR.

If you happen to have one handy, perhaps you could try flashing the T40 with a programmer that sets `pgm->cmd_tpi`? I would wager that would trigger the behavior I'm experiencing. Let me know!

MCUdude commented:
> I'm using an FTDI chip in MPSSE bit-bang mode, which [does seem to set cmd_tpi](https://github.com/avrdudes/avrdude/blob/main/src/avrftdi_tpi.c#L98), meaning that it enters the block in avr_write_mem that I patched in my PR.

I don't have an FT2232-based programmer, but I have several other TPI programmers if that helps.
Can you check if any of the following programmers may be affected?

* [FT232 based  programmer (bit-banging)](https://www.ebay.com/itm/185335203316)
* USBtinyISP
* AVRISPmkII / AVR Dragon

BTW I think we have a spare ATtiny20 at work I can borrow for testing.

danapplegate commented:
@MCUdude 

Yes, I think the FT232 one should work. It looks like it uses [the same underlying configuration](https://github.com/avrdudes/avrdude/blob/44fe5bec2dd2f016d135f0d2f02173045d7aeef5/src/avrdude.conf.in#L887-L923) as the UM232H I'm using. Both are type `"avrftdi"`, which [set `pgm->cmd_tpi`](https://github.com/avrdudes/avrdude/blob/main/src/avrftdi_tpi.c#L98).

Thanks for all your help in tracking this issue down!

MCUdude commented:
Sorry, but I'm not able to reproduce the issue with my current setup:

```
$ ./avrdude -c arduino-ft232r -P /dev/cu.usbserial-A50285BI -pattiny20 -Uflash:w:/Users/hans/Downloads/tpi_dword_test_t10.hex:i

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.04s

avrdude: Device signature = 0x1e910f (probably t20)
avrdude: NOTE: "flash" memory has been specified, an erase cycle will be performed
         To disable this feature, specify the -D option.
avrdude: erasing chip
avrdude: reading input file /Users/hans/Downloads/tpi_dword_test_t10.hex for flash
avrdude: writing 54 bytes flash ...

Writing | ################################################## | 100% 0.23s

avrdude: 54 bytes of flash written
avrdude: verifying flash memory against /Users/hans/Downloads/tpi_dword_test_t10.hex

Reading | ################################################## | 100% 0.12s

avrdude: 54 bytes of flash verified

avrdude done.  Thank you.
```

For reference, here's the verbose output:

<details>

```
$ ./avrdude -c arduino-ft232r -P /dev/cu.usbserial-A50285BI -pattiny20 -Uflash:w:/Users/hans/Downloads/tpi_dword_test_t10.hex:i -vv

avrdude: Version 7.0-20220508
         Copyright (c) Brian Dean, http://www.bdmicro.com/
         Copyright (c) Joerg Wunsch

         System wide configuration file is "/Users/hans/Downloads/avrdude/src/avrdude.conf"
         User configuration file is "/Users/hans/.avrduderc"

         Using Port                    : /dev/cu.usbserial-A50285BI
         Using Programmer              : arduino-ft232r
avrdude: ft245r_open(): no device identifier in portname, using default
ft245r_set_bitclock: bitclk 150000 -> FTDI rate 150000, baud multiplier 1
         AVR Part                      : ATtiny20
         RESET disposition             : dedicated
         RETRY pulse                   : SCK
         Serial program mode           : yes
         Parallel program mode         : yes
         Memory Detail                 :

                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           fuse                    0     0     4    0 no          1   16      0     0     0 0x00 0x00
           lockbits                0     0     0    0 no          1   16      0     0     0 0x00 0x00
           signature               0     0     0    0 no          3   16      0     0     0 0x00 0x00
           calibration             0     0     0    0 no          1   16      0     0     0 0x00 0x00
           flash                   0     0   128    0 no       2048   16      0     0     0 0x00 0x00

         Programmer Type : ftdi_syncbb
         Description     : Arduino: FT232R connected to ISP
         Pin assignment  : 0..7 = DBUS0..7
           VCC     =  (not used)
           BUFF    =  (not used)
           RESET   =  7
           SCK     =  5
           MOSI    =  6
           MISO    =  3
           ERR LED =  (not used)
           RDY LED =  (not used)
           PGM LED =  (not used)
           VFY LED =  (not used)

MOSI-MISO link present
ft245r_cmd_tpi: [ C2 07 ] [ ]
ft245r_cmd_tpi: [ 8F ] [ 80 ]
ft245r_cmd_tpi: [ E0 FF 88 D8 CD 45 AB 89 12 ] [ ]
ft245r_cmd_tpi: [ 80 ] [ 02 ]
avrdude: AVR device initialized and ready to accept instructions

Reading |                                                    | 0% 0.00sft245r_cmd_tpi: [ 72 ] [ 00 ]
ft245r_cmd_tpi: [ F3 00 ] [ ]
ft245r_cmd_tpi: [ 68 C0 ] [ ]
ft245r_cmd_tpi: [ 69 3F ] [ ]
ft245r_cmd_tpi: [ 20 ] [ 1E ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
ft245r_cmd_tpi: [ F3 00 ] [ ]
ft245r_cmd_tpi: [ 68 C1 ] [ ]
ft245r_cmd_tpi: [ 69 3F ] [ ]
ft245r_cmd_tpi: [ 20 ] [ 91 ]
Reading | #################                                  | 33% 0.05sft245r_cmd_tpi: [ 72 ] [ 00 ]
ft245r_cmd_tpi: [ F3 00 ] [ ]
ft245r_cmd_tpi: [ 68 C2 ] [ ]
ft245r_cmd_tpi: [ 69 3F ] [ ]
ft245r_cmd_tpi: [ 20 ] [ 0F ]
Reading | ################################################## | 100% 0.08s

avrdude: Device signature = 0x1e910f (probably t20)
avrdude: NOTE: "flash" memory has been specified, an erase cycle will be performed
         To disable this feature, specify the -D option.
avrdude: erasing chip
ft245r_cmd_tpi: [ 72 ] [ 00 ]
ft245r_cmd_tpi: [ 68 01 69 40 F3 10 60 FF ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
avrdude: reading input file /Users/hans/Downloads/tpi_dword_test_t10.hex for flash
         with 54 bytes in 1 section within [0, 0x35]
         using 4 pages and 10 pad bytes
avrdude: writing 54 bytes flash ...

Writing |                                                    | 0% 0.00sft245r_cmd_tpi: [ 72 ] [ 00 ]
ft245r_cmd_tpi: [ F3 1D ] [ ]
ft245r_cmd_tpi: [ 68 00 ] [ ]
ft245r_cmd_tpi: [ 69 40 ] [ ]
ft245r_cmd_tpi: [ 64 0A ] [ ]
ft245r_cmd_tpi: [ 64 C0 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
ft245r_cmd_tpi: [ 64 11 ] [ ]
ft245r_cmd_tpi: [ 64 C0 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ##                                                 | 3% 0.02sft245r_cmd_tpi: [ 64 10 ] [ ]
ft245r_cmd_tpi: [ 64 C0 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ####                                               | 7% 0.03sft245r_cmd_tpi: [ 64 0F ] [ ]
ft245r_cmd_tpi: [ 64 C0 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ######                                             | 11% 0.04sft245r_cmd_tpi: [ 64 0E ] [ ]
ft245r_cmd_tpi: [ 64 C0 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | #######                                            | 14% 0.05sft245r_cmd_tpi: [ 64 0D ] [ ]
ft245r_cmd_tpi: [ 64 C0 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | #########                                          | 18% 0.06sft245r_cmd_tpi: [ 64 0C ] [ ]
ft245r_cmd_tpi: [ 64 C0 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ###########                                        | 22% 0.06sft245r_cmd_tpi: [ 64 0B ] [ ]
ft245r_cmd_tpi: [ 64 C0 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | #############                                      | 25% 0.08sft245r_cmd_tpi: [ 64 0A ] [ ]
ft245r_cmd_tpi: [ 64 C0 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ###############                                    | 29% 0.08sft245r_cmd_tpi: [ 64 09 ] [ ]
ft245r_cmd_tpi: [ 64 C0 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | #################                                  | 33% 0.10sft245r_cmd_tpi: [ 64 08 ] [ ]
ft245r_cmd_tpi: [ 64 C0 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ###################                                | 37% 0.10sft245r_cmd_tpi: [ 64 11 ] [ ]
ft245r_cmd_tpi: [ 64 27 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ####################                               | 40% 0.11sft245r_cmd_tpi: [ 64 1F ] [ ]
ft245r_cmd_tpi: [ 64 BF ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ######################                             | 44% 0.11sft245r_cmd_tpi: [ 64 CF ] [ ]
ft245r_cmd_tpi: [ 64 E5 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ########################                           | 48% 0.13sft245r_cmd_tpi: [ 64 D0 ] [ ]
ft245r_cmd_tpi: [ 64 E0 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ##########################                         | 51% 0.13sft245r_cmd_tpi: [ 64 DE ] [ ]
ft245r_cmd_tpi: [ 64 BF ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ############################                       | 55% 0.14sft245r_cmd_tpi: [ 64 CD ] [ ]
ft245r_cmd_tpi: [ 64 BF ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ##############################                     | 59% 0.15sft245r_cmd_tpi: [ 64 02 ] [ ]
ft245r_cmd_tpi: [ 64 D0 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ###############################                    | 62% 0.16sft245r_cmd_tpi: [ 64 06 ] [ ]
ft245r_cmd_tpi: [ 64 C0 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | #################################                  | 66% 0.16sft245r_cmd_tpi: [ 64 EC ] [ ]
ft245r_cmd_tpi: [ 64 CF ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ###################################                | 70% 0.17sft245r_cmd_tpi: [ 64 00 ] [ ]
ft245r_cmd_tpi: [ 64 00 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | #####################################              | 74% 0.18sft245r_cmd_tpi: [ 64 00 ] [ ]
ft245r_cmd_tpi: [ 64 00 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | #######################################            | 77% 0.19sft245r_cmd_tpi: [ 64 90 ] [ ]
ft245r_cmd_tpi: [ 64 E0 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | #########################################          | 81% 0.19sft245r_cmd_tpi: [ 64 80 ] [ ]
ft245r_cmd_tpi: [ 64 E0 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ###########################################        | 85% 0.20sft245r_cmd_tpi: [ 64 08 ] [ ]
ft245r_cmd_tpi: [ 64 95 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ############################################       | 88% 0.21sft245r_cmd_tpi: [ 64 F8 ] [ ]
ft245r_cmd_tpi: [ 64 94 ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ##############################################     | 92% 0.22sft245r_cmd_tpi: [ 64 FF ] [ ]
ft245r_cmd_tpi: [ 64 CF ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ################################################   | 96% 0.22sft245r_cmd_tpi: [ 64 FF ] [ ]
ft245r_cmd_tpi: [ 64 FF ] [ ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 80 ]
ft245r_cmd_tpi: [ 72 ] [ 00 ]
Writing | ################################################## | 100% 0.23s

avrdude: 54 bytes of flash written
avrdude: verifying flash memory against /Users/hans/Downloads/tpi_dword_test_t10.hex
avrdude: reading on-chip flash data ...

Reading |                                                    | 0% 0.00sft245r_cmd_tpi: [ 72 ] [ 00 ]
ft245r_cmd_tpi: [ F3 00 ] [ ]
ft245r_cmd_tpi: [ 68 00 ] [ ]
ft245r_cmd_tpi: [ 69 40 ] [ ]
ft245r_cmd_tpi: [ 24 ] [ 0A ]
ft245r_cmd_tpi: [ 24 ] [ C0 ]
ft245r_cmd_tpi: [ 24 ] [ 11 ]
ft245r_cmd_tpi: [ 24 ] [ C0 ]
ft245r_cmd_tpi: [ 24 ] [ 10 ]
ft245r_cmd_tpi: [ 24 ] [ C0 ]
ft245r_cmd_tpi: [ 24 ] [ 0F ]
ft245r_cmd_tpi: [ 24 ] [ C0 ]
ft245r_cmd_tpi: [ 24 ] [ 0E ]
ft245r_cmd_tpi: [ 24 ] [ C0 ]
ft245r_cmd_tpi: [ 24 ] [ 0D ]
ft245r_cmd_tpi: [ 24 ] [ C0 ]
ft245r_cmd_tpi: [ 24 ] [ 0C ]
ft245r_cmd_tpi: [ 24 ] [ C0 ]
ft245r_cmd_tpi: [ 24 ] [ 0B ]
ft245r_cmd_tpi: [ 24 ] [ C0 ]
ft245r_cmd_tpi: [ 24 ] [ 0A ]
ft245r_cmd_tpi: [ 24 ] [ C0 ]
ft245r_cmd_tpi: [ 24 ] [ 09 ]
ft245r_cmd_tpi: [ 24 ] [ C0 ]
ft245r_cmd_tpi: [ 24 ] [ 08 ]
ft245r_cmd_tpi: [ 24 ] [ C0 ]
Reading | #                                                  | 1% 0.06sft245r_cmd_tpi: [ 24 ] [ 11 ]
ft245r_cmd_tpi: [ 24 ] [ 27 ]
ft245r_cmd_tpi: [ 24 ] [ 1F ]
ft245r_cmd_tpi: [ 24 ] [ BF ]
ft245r_cmd_tpi: [ 24 ] [ CF ]
ft245r_cmd_tpi: [ 24 ] [ E5 ]
ft245r_cmd_tpi: [ 24 ] [ D0 ]
ft245r_cmd_tpi: [ 24 ] [ E0 ]
ft245r_cmd_tpi: [ 24 ] [ DE ]
ft245r_cmd_tpi: [ 24 ] [ BF ]
ft245r_cmd_tpi: [ 24 ] [ CD ]
ft245r_cmd_tpi: [ 24 ] [ BF ]
ft245r_cmd_tpi: [ 24 ] [ 02 ]
ft245r_cmd_tpi: [ 24 ] [ D0 ]
ft245r_cmd_tpi: [ 24 ] [ 06 ]
ft245r_cmd_tpi: [ 24 ] [ C0 ]
ft245r_cmd_tpi: [ 24 ] [ EC ]
ft245r_cmd_tpi: [ 24 ] [ CF ]
ft245r_cmd_tpi: [ 24 ] [ 00 ]
ft245r_cmd_tpi: [ 24 ] [ 00 ]
Reading | #                                                  | 2% 0.12sft245r_cmd_tpi: [ 24 ] [ 00 ]
ft245r_cmd_tpi: [ 24 ] [ 00 ]
ft245r_cmd_tpi: [ 24 ] [ 90 ]
ft245r_cmd_tpi: [ 24 ] [ E0 ]
ft245r_cmd_tpi: [ 24 ] [ 80 ]
ft245r_cmd_tpi: [ 24 ] [ E0 ]
ft245r_cmd_tpi: [ 24 ] [ 08 ]
ft245r_cmd_tpi: [ 24 ] [ 95 ]
ft245r_cmd_tpi: [ 24 ] [ F8 ]
ft245r_cmd_tpi: [ 24 ] [ 94 ]
ft245r_cmd_tpi: [ 24 ] [ FF ]
ft245r_cmd_tpi: [ 24 ] [ CF ]
Reading | ################################################## | 100% 0.15s

avrdude: verifying ...
avrdude: 54 bytes of flash verified

avrdude done.  Thank you.

```

</details>

danapplegate commented:
🎉 

Whew! Glad I wasn't going crazy.

In regards to the T40, I think it only worked because our test program has 27 words, and with my patch writing an extra word to bring it to 28, that happens to be divisible by four. I bet if we try a program that has, say 26 words, it would still fail.

I can take another pass at the fix to see if we can solve it for the T40 as well. Since this only seems to affect the T20 and T40, I'm thinking about just matching the part id, rather than a full config option. What do you think?

MCUdude commented:
>

Yup, can confirm that word sizes that are dividable by four is an issue on ATtiny40:

```
$ ./avrdude -c arduino-ft232r -P /dev/cu.usbserial-A50285BI -pattiny40 -Uflash:w:0xaa_20words.hex:i

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.03s

avrdude: Device signature = 0x1e920e (probably t40)
avrdude: NOTE: "flash" memory has been specified, an erase cycle will be performed
         To disable this feature, specify the -D option.
avrdude: erasing chip
avrdude: reading input file 0xaa_20words.hex for flash
avrdude: writing 50 bytes flash ...

Writing | ################################################## | 100% 0.17s

avrdude: 50 bytes of flash written
avrdude: verifying flash memory against 0xaa_20words.hex

Reading | ################################################## | 100% 0.11s

avrdude: verification error, first mismatch at byte 0x0030
         0xff != 0xaa
avrdude: verification error; content mismatch

avrdude done.  Thank you.
```

> I can take another pass at the fix to see if we can solve it for the T40 as well. Since this only seems to affect the T20 and T40, I'm thinking about just matching the part id, rather than a full config option. What do you think?

We have been trying our best to not write code that's hardwired to specific targets, but instead use a property in avrdude.conf or similar. However, padding the last word with 0xff should be harmless. I'd be interested to hear what @stefanrueger has to say!

stefanrueger commented:
@danapplegate  and @MCUdude Thanks for careful analysis. Two comments:
1. I would defo make this dependent on a (new) `avrdude.conf` memory property. Can I suggest to use `n_word_writes` or something to that effect and implement similar to `n_page_erase`: In `config.c` add the following lines to the end of `avr_comp[]` array
```
  // AVRMEM
  part_mem_desc(n_word_writes, COMP_INT),
```
Also add corresponding lines in `developer_opts.c`, `libavrdude.h` (both, eg, below `num_pages`) and change `doc/avrdude.texi` and `avrdude.conf.in` correspondingly

2. I don't think this PR actually works for `t20` in the general case of input files with holes in them: for example, users are free to construct their flash `.hex` files in program and sections as they like. These sections can be non-contiguous, contain an odd number of bytes, etc. Sooo, I'd first figure out whether 2 bytes (`n_word_writes` not set), 4 bytes (`n_word_writes` is 2, set in flash section of t20) or 8 bytes (`n_word_writes` is 4) need to be treated at the same time
```
  if(m->n_word_writes < 0 || m->n_word_writes > 4 || m->n_word_writes == 3) {
    avrdude_message(MSG_INFO, ...); // complain and leave
    return/exit/goto...
  }
  int chunk = m->n_word_writes>0? 2*m->n_word_writes: 2;
```
Then I'd ensure wsize is a multiple of chunk
```
  wsize = (wsize+chunk-1)/chunk*chunk;
```
Then loop through memory from 0 to wsize in steps of chunk
  - See whether one of the bytes in this chunk region was set by input (one of the i's has `m->tags[i] & TAG_ALLOCATED`)
  - If so write this chunk to the device using the correct method (2, 4 ot 8 bytes to write);
  - I think(!)  holes in the input file appear as 0xff here, so not necessary to pad holes with 0xff; as flash for these devices looks like NOR-memory writing 0xff is a nop, so all good (otherwise I suggest byte-wise reading holes from device to pad) 
 
This to replace the current loop, lines 874ff in avr.c

Voila! 

PS: @MCUdude and @mariusgreuel: can you check/verify whether this PR, once done, is a suitable generalisation of PR #825? Somehow that latter kind of fell between the cracks.

MCUdude commented:
> PS: MCUdude and mariusgreuel: can you check/verify whether this PR, once done, is a suitable generalisation of PR https://github.com/avrdudes/avrdude/pull/825? Somehow that latter kind of fell between the cracks.

Sure! Once a general solution is ready, I can try it on all TPI devices I have access to (T10, T20, T40, T104). Just tell me how many bytes to write to each target, and I'll provide some test results!

stefanrueger commented:
>  I can try it on all TPI devices I have access to (T10, T20, T40, T104).

Brill! Perhaps generate a difficult test .hex file with holes in them using srec_cat or similar?

MCUdude commented:
> I would defo make this dependent on a (new) avrdude.conf memory property. Can I suggest to use n_word_writes or something to that effect and implement similar to n_page_erase: In config.c add the following lines to the end of avr_comp[] array

I think it's a good idea to base this "fix" on actual data provided in avrdude.conf. Just bear in mind that it has to be an optional parameter to maintain backwards compatibility with older avrdude.conf files.

> Brill! Perhaps generate a difficult test .hex file with holes in them using srec_cat or similar?

Sure! I've only been scratching the surface of what srec_cat can do, and only generated simple hex files, without holes, that repeats the same byte over and over again. Do you have a command example on how I can generate a file with one or more holes?

stefanrueger commented:
> optional parameter 

Sure, when not preset or not set should behave as without this PR.

> command example

How about the following bash script. Assumes you have a working blink.hex for that device.
```
yes "Bonjour le monde" | head -41c | srec_cat - -raw -offset 0x456 -o - -intel >/tmp/1
yes "Hello, world"     | head -23c | srec_cat - -raw -offset 0x567 -o - -intel >/tmp/2
yes "Hei Verden"       | head -71c | srec_cat - -raw -offset 0x678 -o - -intel >/tmp/3
srec_cat blink.hex -intel /tmp/1 -intel /tmp/2 -intel /tmp/3 -intel -o - -intel >blink+holes+data.hex
rm /tmp/1 /tmp/2 /tmp/3
```

danapplegate commented:
@stefanrueger Thanks for the guidance! I took your suggestions and added the `n_word_writes` config option and modified the TPI flashing loop to write words in chunks. Please take a look and let me know if you have any feedback.

@MCUdude I've tested these changes on my ATtiny20 with a 33 word contiguous hex file, and this version still seems to fix the issue. If you are able to test with the T40 and with some non-contiguous hex files, that would be awesome!

stefanrueger commented:
@danapplegate Brilliant! Thanks for changes. I have pushed a small commit for using a simplified way to add a new component to the grammar that does not need changes to config_gram.y. Once @MCUdude has reviewed the PR and confirmed the PR working correctly, I am happy to squash and merge this one into git mainline.

MCUdude commented:
Sorry, but I'm not able to communicate with either the ATtiny20 or the ATtiny40. If I build from git main, I'm at least able to read the device signature:

```
$ ./avrdude -c arduino-ft232r -P /dev/cu.usbserial-A50285BI -pattiny20

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.03s

avrdude: Device signature = 0x000000 (retrying)

Reading | ################################################## | 100% 0.03s

avrdude: Device signature = 0x000000 (retrying)

Reading | ################################################## | 100% 0.03s

avrdude: Device signature = 0x000000
avrdude: Yikes!  Invalid device signature.
         Double check connections and try again, or use -F to override
         this check.


avrdude done.  Thank you.
```

<p>

```
$ ./avrdude -c arduino-ft232r -P /dev/cu.usbserial-A50285BI -pattiny40

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.03s

avrdude: Device signature = 0x000000 (retrying)

Reading | ################################################## | 100% 0.04s

avrdude: Device signature = 0x000000 (retrying)

Reading | ################################################## | 100% 0.03s

avrdude: Device signature = 0x000000
avrdude: Yikes!  Invalid device signature.
         Double check connections and try again, or use -F to override
         this check.


avrdude done.  Thank you.
```

MCUdude commented:
The issue seems to be caused by commit a8848174f3173d685191c3e17b2526b9c3001cb4.

MCUdude commented:
I don't see any obvious issues in https://github.com/avrdudes/avrdude/commit/a8848174f3173d685191c3e17b2526b9c3001cb4.

What also strange is that I'm getting av avrdude.conf warning (ATmega2560), when I'm reading 0x000000 as device signature (commit https://github.com/avrdudes/avrdude/commit/a8848174f3173d685191c3e17b2526b9c3001cb4), but not with commit 087243df9d2d2c52dd4930431bc39eb5dc1deb62. See output logs below:

```
$ git branch
* (HEAD detached at a884817)

$ ./avrdude -c arduino-ft232r -P /dev/cu.usbserial-A50285BI -pattiny20
avrdude: warning at /Users/hans/Downloads/avrdude-terminal/src/avrdude.conf:9071: ATmega2560's flash load_ext_addr misses a necessary address bit a16
avrdude: warning at /Users/hans/Downloads/avrdude-terminal/src/avrdude.conf:9071: ATmega2560's flash writepage misses a necessary address bit a15

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.03s

avrdude: Device signature = 0x000000 (retrying)

Reading | ################################################## | 100% 0.03s

avrdude: Device signature = 0x000000 (retrying)

Reading | ################################################## | 100% 0.03s

avrdude: Device signature = 0x000000
avrdude: Yikes!  Invalid device signature.
         Double check connections and try again, or use -F to override
         this check.


avrdude done.  Thank you.
```

```
$ git branch
* (HEAD detached at 087243d)

$ ./avrdude -c arduino-ft232r -P /dev/cu.usbserial-A50285BI -pattiny20

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.04s

avrdude: Device signature = 0x1e910f (probably t20)

avrdude done.  Thank you.
```

danapplegate commented:
@MCUdude Hmm, very strange. Can you try backing up your current avrdude.conf file and rebuilding to generate a fresh one, and see if that fixes the issue?

stefanrueger commented:
@MCUdude I suspect a problem in the workflow/Makefile/CMakeList.txt ... Something along the lines that switching between branches does not force a rebuild of the things that were changed (make uses date, cmake hashes to identify changes). So you may be linking .o files from different branches, always a recipes for disaster and odd error messages.

MCUdude commented:
> @MCUdude I suspect a problem in the workflow/Makefile/CMakeList.txt ... Something along the lines that switching between branches does not force a rebuild of the things that were changed (make uses date, cmake hashes to identify changes). So you may be linking .o files from different branches, always a recipes for disaster and odd error messages.

This was probably the issue. I use make, and after running `make clean`, it finally started working again.

It appears to work on T10, T20, T40 and T104.

```
$ ./avrdude -c arduino-ft232r -P /dev/cu.usbserial-A50285BI -pattiny10 -Uflash:w:nop+holes+data.hex:i

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.04s

avrdude: Device signature = 0x1e9003 (probably t10)
avrdude: NOTE: "flash" memory has been specified, an erase cycle will be performed
         To disable this feature, specify the -D option.
avrdude: erasing chip
avrdude: reading input file nop+holes+data.hex for flash
         with 189 bytes in 4 sections within [0, 0x38b]
         using 14 pages and 35 pad bytes
avrdude: writing 189 bytes flash ...

Writing | ################################################## | 100% 0.95s

avrdude: 189 bytes of flash written
avrdude: verifying flash memory against nop+holes+data.hex

Reading | ################################################## | 100% 0.43s

avrdude: 189 bytes of flash verified

avrdude done.  Thank you.


$ ./avrdude -c arduino-ft232r -P /dev/cu.usbserial-A50285BI -pattiny20 -Uflash:w:nop+holes+data.hex:i

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.04s

avrdude: Device signature = 0x1e910f (probably t20)
avrdude: NOTE: "flash" memory has been specified, an erase cycle will be performed
         To disable this feature, specify the -D option.
avrdude: erasing chip
avrdude: reading input file nop+holes+data.hex for flash
         with 189 bytes in 4 sections within [0, 0x38b]
         using 14 pages and 35 pad bytes
avrdude: writing 189 bytes flash ...

Writing | ################################################## | 100% 0.72s

avrdude: 189 bytes of flash written
avrdude: verifying flash memory against nop+holes+data.hex

Reading | ################################################## | 100% 0.43s

avrdude: 189 bytes of flash verified

avrdude done.  Thank you.


$ ./avrdude -c arduino-ft232r -P /dev/cu.usbserial-A50285BI -pattiny40 -Uflash:w:nop+holes+data.hex:i

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.03s

avrdude: Device signature = 0x1e920e (probably t40)
avrdude: NOTE: "flash" memory has been specified, an erase cycle will be performed
         To disable this feature, specify the -D option.
avrdude: erasing chip
avrdude: reading input file nop+holes+data.hex for flash
         with 189 bytes in 4 sections within [0, 0x38b]
         using 7 pages and 259 pad bytes
avrdude: writing 189 bytes flash ...

Writing | ################################################## | 100% 0.57s

avrdude: 189 bytes of flash written
avrdude: verifying flash memory against nop+holes+data.hex

Reading | ################################################## | 100% 0.41s

avrdude: 189 bytes of flash verified

avrdude done.  Thank you.


$ ./avrdude -c arduino-ft232r -P /dev/cu.usbserial-A50285BI -pattiny104 -Uflash:w:nop+holes+data.hex:i

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.03s

avrdude: Device signature = 0x1e900b (probably t104)
avrdude: NOTE: "flash" memory has been specified, an erase cycle will be performed
         To disable this feature, specify the -D option.
avrdude: erasing chip
avrdude: reading input file nop+holes+data.hex for flash
         with 189 bytes in 4 sections within [0, 0x38b]
         using 14 pages and 35 pad bytes
avrdude: writing 189 bytes flash ...

Writing | ################################################## | 100% 0.91s

avrdude: 189 bytes of flash written
avrdude: verifying flash memory against nop+holes+data.hex

Reading | ################################################## | 100% 0.41s

avrdude: 189 bytes of flash verified

avrdude done.  Thank you.


```

danapplegate commented:
@stefanrueger It looks like Hans has been able to test and confirm the fix for a variety of ATtiny devices. Is there any other testing to be done before merging, do you think?

stefanrueger commented:
@danapplegate Plan is to merge this PR with others (a batch is easier to handle) soonish, also to give @mariusgreuel a sporting change to review it. @MCUdude Thankd for testing, well appreciated. Could you make the used test file with holes available to others for their tests, please?

MCUdude commented:
> Could you make the used test file with holes available to others for their tests, please?

Sure, but it's nothing special. Just the test program @danapplegate provided, with additional holes. Also slightly tweaked in order to fit on a 1 kiB chip.

[nop+holes+data.hex.zip](https://github.com/avrdudes/avrdude/files/9780923/nop%2Bholes%2Bdata.hex.zip)
